sdk/metric/aggregator/exponential/structure/exponential.go

```diff
@@ -0,0 +1,657 @@
+	cfg, _ = cfg.Validate()
+
+	h.maxSize = cfg.maxSize
+
+	mapping, _ := newMapping(logarithm.MaxScale)
+	h.mapping = mapping
+}
+
+// Sum implements aggregation.Histogram.
+func (h *Histogram[N]) Sum() N {
+	return h.sum
+}
+
+// Min implements aggregation.Histogram.
+func (h *Histogram[N]) Min() N {
+	return h.min
+}
+
+// Max implements aggregation.Histogram.
+func (h *Histogram[N]) Max() N {
+	return h.max
+}
+
+// Count implements aggregation.Histogram.
+func (h *Histogram[N]) Count() uint64 {
+	return h.count
+}
+
+// Scale implements aggregation.Histogram.
+func (h *Histogram[N]) Scale() int32 {
+	if h.count == h.zeroCount {
+		// all zeros! scale doesn't matter, use zero.
+		return 0
+	}
+	return h.mapping.Scale()
+}
+
+// ZeroCount implements aggregation.Histogram.
+func (h *Histogram[N]) ZeroCount() uint64 {
+	return h.zeroCount
+}
+
+// Positive implements aggregation.Histogram.
+func (h *Histogram[N]) Positive() *Buckets {
+	return &h.positive
+}
+
+// Negative implements aggregation.Histogram.
+func (h *Histogram[N]) Negative() *Buckets {
+	return &h.negative
+}
+
+// Offset implements aggregation.Bucket.
+func (b *Buckets) Offset() int32 {
+	return b.indexStart
+}
+
+// Len implements aggregation.Bucket.
+func (b *Buckets) Len() uint32 {
+	if b.backing == nil {
+		return 0
+	}
+	if b.indexEnd == b.indexStart && b.At(0) == 0 {
+		return 0
+	}
+	return uint32(b.indexEnd - b.indexStart + 1)
+}
+
+// At returns the count of the bucket at a position in the logical
+// array of counts.
+func (b *Buckets) At(pos0 uint32) uint64 {
+	pos := pos0
+	bias := uint32(b.indexBase - b.indexStart)
+
+	if pos < bias {
+		pos += uint32(b.backing.size())
+	}
+	pos -= bias
+
+	return b.backing.countAt(pos)
+}
+
+// Clear resets a histogram to the empty state without changing
+// backing array.
+func (h *Histogram[N]) Clear() {
+	h.positive.clear()
+	h.negative.clear()
+	h.sum = 0
+	h.count = 0
+	h.zeroCount = 0
+	h.min = 0
+	h.max = 0
+	h.mapping, _ = newMapping(logarithm.MaxScale)
+}
+
+// clear zeros the backing array.
+func (b *Buckets) clear() {
+	b.indexStart = 0
+	b.indexEnd = 0
+	b.indexBase = 0
+	if b.backing != nil {
+		b.backing.reset()
+	}
+}
+
+func newMapping(scale int32) (mapping.Mapping, error) {
+	if scale <= 0 {
+		return exponent.NewMapping(scale)
+	}
+	return logarithm.NewMapping(scale)
+}
+
+// Swap exchanges the contents of `h` and `dest`.
+func (h *Histogram[N]) Swap(dest *Histogram[N]) {
+	*dest, *h = *h, *dest
+}
+
+// CopyInto copies `h` into `dest`.
+func (h *Histogram[N]) CopyInto(dest *Histogram[N]) {
+	dest.Clear()
+	dest.MergeFrom(h)
+}
+
+// Update supports updating a histogram with a single count.
+func (h *Histogram[N]) Update(number N) {
+	h.UpdateByIncr(number, 1)
+}
+
+// UpdateByIncr supports updating a histogram with a non-negative
+// increment.
+func (h *Histogram[N]) UpdateByIncr(number N, incr uint64) {
+	value := float64(number)
+
+	// Maintain min and max
+	if h.count == 0 {
```

Comment:
if seems redundant when count == 0, the else part should still correctly set the min and max?